hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/KafkaReadFunction.java

```diff
@@ -67,59 +69,99 @@ public KafkaReadFunction(OperationContext context, String topicName, String boot
+                        LOG.warn( id+" KRF.call kafkaRecords Spark TaskContext Interrupted");
+                        //consumer.close();
                         throw new TaskKilledException();
                     }
                 } while( noRecords.test(records) && attempt++ < maxAttempts );
                 
                 return records;
             }
             
-            private boolean hasMoreRecords(int maxAttempts) throws TaskKilledException {
-                ConsumerRecords<Integer, Externalizable> records = kafkaRecords(maxAttempts);
+            private boolean hasMoreRecords(int maxAttempts, Duration timeout) throws TaskKilledException {
+                ConsumerRecords<Integer, byte[]> records = kafkaRecords(maxAttempts, timeout);
                 if( noRecords.test(records) ) {
-                    consumer.close();
+                    if( !prevMessage.last() && retries < maxRetries ) {
+                        retries++;
+                        Duration retryTimeout = longTimeout;
+                        LOG.warn( id+" KRF.call Missed rcds, got "+totalCount+" retry "+retries+" for up to "+retryTimeout );
+                        return hasMoreRecords(
+                            maxAttempts,
+                            retryTimeout
+                        );
+                    }
+                    //consumer.close();
+                    if( !prevMessage.last() ) {
+                        LOG.error(id + " KRF.call Didn't get full batch after " + retries + " retries, got " + totalCount + " records");
+                    }
                     return false;
                 } else {
+                    int ct = records.count();
+                    totalCount += ct;
+                    LOG.trace( id+" KRF.call p "+partition+" t "+topicName+" records "+ct );
+                    retries = 0;
+                    
                     it = records.iterator();
                     return it.hasNext();
                 }
             }
-
+            
             @Override
             public boolean hasNext() {
-                if (it == null) {
-                    return hasMoreRecords(60);
-                }
-                if (it.hasNext()) {
-                    return true;
+                boolean more = false;
+                
+                if (it != null && it.hasNext()) {
+                    more = true;
+                } else if (!prevMessage.last()) {
+                    more = hasMoreRecords(1, shortTimeout);
                 }
-                else {
-                    return hasMoreRecords(1);
+
+                if (!more) {
+                    consumer.close();
+                    kryo.close();
                 }
+
+                return more;
             }
 
             @Override
             public ExecRow next() {
-                return (ExecRow)it.next().value();
+                Message m = (Message)kryo.deserialize( it.next().value() );
```

Comment:
Serializing ValueRows directly with Kryo still has quite a bit of overhead (we are serializing the schema with each row, for instance). I opened https://splicemachine.atlassian.net/browse/DB-10543 to track this improvement